"colon" violation no longer autocorrected in dictionary literals (since 0.23.1)

### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [X] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Bug Report

`swiftlint autocorrect` does not correct the `colon` rule on some cases (especially in dictionary literals?).

I've tested with versions 0.24.0 thru 0.25.0 but the bug is still present, while it was working in 0.23.1

### Environment

* SwiftLint versions: 0.23.1 OK, but bug present with 0.24.0, 0.24.1, 0.24.2 and 0.25.0
* Installation method used: portable zip (also tested 0.23.1 with homebrew, and 0.25.0 with CocoaPods)
* Which Xcode version are you using (check `xcode-select -p`)? 9.2


### Sample code to reproduce

#### Sample swift file

```swift
import UIKit

class SwiftLintBugSample: UIViewController {

    fileprivate func updateTableView() {

        let dict = [KeysConstants.CELL_KEYS.CELL_TYPE : FooCellType.editTextCell.rawValue,
                    KeysConstants.CELL_KEYS.CELL_IDENTIFIER_KEY : FooEditTextTableViewCell.cellIdentifier,
                    KeysConstants.CELL_KEYS.CELL_GRAY : false,
                    KeysConstants.CELL_KEYS.VALUE_ERROR : self.isInError,
                    KeysConstants.CELL_KEYS.ESTIMATED_CELL_HEIGHT : FooEditTextTableViewCell.cellHeight] as [String: Any]

        self.dataSource = dict as NSDictionary

    }
}
```

#### Configuration File

```yml
whitelist_rules:
  - colon
```

#### Test script

This script aims to test a given swiftlint version by linting first to show the colon violations, then trying to autocorrect them, then linting again to check if they have indeed been corrected.

```sh
#!/bin/sh

bin=$1/swiftlint
echo ">>> Testing with version `$bin version`"

git checkout sample.swift

echo "------ Initial linting ------"
$bin lint --path sample.swift

echo "------ Autocorrecting  ------"
$bin autocorrect --path sample.swift

echo "------  Linting again  ------"
$bin lint --path sample.swift
```

> Note: I made this test script in order to test the bug across multiple versions of SwiftLint. In order for this script to work, I've downloaded various versions of SwiftLint (portable zip) and placed each version in its own folder (each folder named after the corresponding version)

#### Output for various Swiftlint versions

<details>
  <summary>swiftlint 0.23.1</summary>

> This version works as expected: 5 violations before autocorrect, all corrected, no more violation after autocorrect.

  ```sh
$ ./test.sh 0.23.1
>>> Testing with version 0.23.1
------ Initial linting ------
Loading configuration from '.swiftlint.yml'
Linting Swift files at path sample.swift
Linting 'sample.swift' (1/1)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:7:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:8:64: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:9:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:10:56: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:11:66: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
Done linting! Found 5 violations, 0 serious in 1 file.
------ Autocorrecting  ------
Loading configuration from '.swiftlint.yml'
Correcting Swift files at path sample.swift
Correcting 'sample.swift' (1/1)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:11:66 Corrected Colon
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:10:56 Corrected Colon
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:9:54 Corrected Colon
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:8:64 Corrected Colon
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:7:54 Corrected Colon
Done correcting 1 files!
------  Linting again  ------
Loading configuration from '.swiftlint.yml'
Linting Swift files at path sample.swift
Linting 'sample.swift' (1/1)
Done linting! Found 0 violations, 0 serious in 1 file.
  ```
</details>

<details>
  <summary>swiftlint 0.24.0</summary>

> This version does not work as expected: 5 violations before autocorrect, none auto-corrected, still 5 violation present after autocorrect.

  ```sh
$ ./test.sh 0.24.0
>>> Testing with version 0.24.0
------ Initial linting ------
Loading configuration from '.swiftlint.yml'
Linting Swift files at path sample.swift
Linting 'sample.swift' (1/1)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:7:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:8:64: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:9:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:10:56: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:11:66: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
Done linting! Found 5 violations, 0 serious in 1 file.
------ Autocorrecting  ------
Loading configuration from '.swiftlint.yml'
Correcting Swift files at path sample.swift
Correcting 'sample.swift' (1/1)
Done correcting 1 files!
------  Linting again  ------
Loading configuration from '.swiftlint.yml'
Linting Swift files at path sample.swift
Linting 'sample.swift' (1/1)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:7:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:8:64: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:9:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:10:56: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:11:66: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
Done linting! Found 5 violations, 0 serious in 1 file.
  ```
</details>

<details>
  <summary>swiftlint 0.24.1</summary>

> This version does not work as expected: 5 violations before autocorrect, none auto-corrected, still 5 violation present after autocorrect.

  ```sh
$ ./test.sh 0.24.1
>>> Testing with version 0.24.1
------ Initial linting ------
Loading configuration from '.swiftlint.yml'
Invalid configuration for 'indentation'. Falling back to default.
Linting Swift files at path sample.swift
Linting 'sample.swift' (1/1)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:7:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:8:64: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:9:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:10:56: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:11:66: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
Done linting! Found 5 violations, 0 serious in 1 file.
------ Autocorrecting  ------
Loading configuration from '.swiftlint.yml'
Invalid configuration for 'indentation'. Falling back to default.
Correcting Swift files at path sample.swift
Correcting 'sample.swift' (1/1)
Done correcting 1 files!
------  Linting again  ------
Loading configuration from '.swiftlint.yml'
Invalid configuration for 'indentation'. Falling back to default.
Linting Swift files at path sample.swift
Linting 'sample.swift' (1/1)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:7:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:8:64: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:9:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:10:56: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:11:66: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
Done linting! Found 5 violations, 0 serious in 1 file.
  ```
</details>

<details>
  <summary>swiftlint 0.24.2</summary>

> This version does not work as expected: 5 violations before autocorrect, none auto-corrected, still 5 violation present after autocorrect.

  ```sh
$ ./test.sh 0.24.2
>>> Testing with version 0.24.2
------ Initial linting ------
Loading configuration from '.swiftlint.yml'
Linting Swift files at path sample.swift
Linting 'sample.swift' (1/1)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:7:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:8:64: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:9:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:10:56: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:11:66: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
Done linting! Found 5 violations, 0 serious in 1 file.
------ Autocorrecting  ------
Loading configuration from '.swiftlint.yml'
Correcting Swift files at path sample.swift
Correcting 'sample.swift' (1/1)
Done correcting 1 files!
------  Linting again  ------
Loading configuration from '.swiftlint.yml'
Linting Swift files at path sample.swift
Linting 'sample.swift' (1/1)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:7:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:8:64: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:9:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:10:56: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:11:66: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
Done linting! Found 5 violations, 0 serious in 1 file.
  ```
</details>

<details>
  <summary>swiftlint 0.25.0</summary>

> This version does not work as expected: 5 violations before autocorrect, none auto-corrected, still 5 violation present after autocorrect.

  ```sh
$ ./test.sh 0.25.0
>>> Testing with version 0.25.0
------ Initial linting ------
Loading configuration from '.swiftlint.yml'
Linting Swift files at path sample.swift
Linting 'sample.swift' (1/1)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:7:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:8:64: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:9:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:10:56: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:11:66: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
Done linting! Found 5 violations, 0 serious in 1 file.
------ Autocorrecting  ------
Loading configuration from '.swiftlint.yml'
Correcting Swift files at path sample.swift
Correcting 'sample.swift' (1/1)
Done correcting 1 files!
------  Linting again  ------
Loading configuration from '.swiftlint.yml'
Linting Swift files at path sample.swift
Linting 'sample.swift' (1/1)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:7:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:8:64: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:9:54: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:10:56: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
/Users/ohalligon/Desktop/swiftlint-bug-test/swiftlint/sample.swift:11:66: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
Done linting! Found 5 violations, 0 serious in 1 file.
  ```
</details>

---

\cc @marcelofabri as discussed on Slack.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,7 +10,9 @@
 
 #### Bug Fixes
 
-* None.
+* Fix some cases where `colon` rule wouldn't autocorrect dictionary literals.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [#2050](https://github.com/realm/SwiftLint/issues/2050)
 
 ## 0.25.0: Cleaning the Lint Filter
 

--- a/Rules.md
+++ b/Rules.md
@@ -1180,6 +1180,10 @@ func abc(def: Void) { ghi(jkl: mno) }
 class ABC { let def = ghi(jkl: mno) } }
 ```
 
+```swift
+func foo() { let dict = [1: 1] }
+```
+
 </details>
 <details>
 <summary>Triggering Examples</summary>
@@ -1414,6 +1418,10 @@ func abc(def: Void) { ghi(jkl↓:mno) }
 
 ```swift
 class ABC { let def = ghi(jkl↓:mno) } }
+```
+
+```swift
+func foo() { let dict = [1↓ : 1] }
 ```
 
 </details>

--- a/Source/SwiftLintFramework/Rules/ColonRule+Dictionary.swift
+++ b/Source/SwiftLintFramework/Rules/ColonRule+Dictionary.swift
@@ -18,12 +18,13 @@ extension ColonRule {
         }
 
         return dictionary.substructure.flatMap { subDict -> [NSRange] in
-            guard let kindString = subDict.kind,
-                let kind = KindType(rawValue: kindString) else {
-                    return []
+            var ranges: [NSRange] = []
+            if let kind = subDict.kind.flatMap(KindType.init(rawValue:)) {
+                ranges += dictionaryColonViolationRanges(in: file, kind: kind, dictionary: subDict)
             }
-            return dictionaryColonViolationRanges(in: file, dictionary: subDict) +
-                dictionaryColonViolationRanges(in: file, kind: kind, dictionary: subDict)
+            ranges += dictionaryColonViolationRanges(in: file, dictionary: subDict)
+
+            return ranges
         }
     }
 

--- a/Source/SwiftLintFramework/Rules/ColonRule.swift
+++ b/Source/SwiftLintFramework/Rules/ColonRule.swift
@@ -61,7 +61,8 @@ public struct ColonRule: CorrectableRule, ConfigurationProviderRule {
             "object.method(5, y: \"string\")\n",
             "func abc() { def(ghi: jkl) }",
             "func abc(def: Void) { ghi(jkl: mno) }",
-            "class ABC { let def = ghi(jkl: mno) } }"
+            "class ABC { let def = ghi(jkl: mno) } }",
+            "func foo() { let dict = [1: 1] }"
         ],
         triggeringExamples: [
             "let ↓abc:Void\n",
@@ -110,7 +111,8 @@ public struct ColonRule: CorrectableRule, ConfigurationProviderRule {
             "object.method(x↓:  5, y: \"string\")\n",
             "func abc() { def(ghi↓:jkl) }",
             "func abc(def: Void) { ghi(jkl↓:mno) }",
-            "class ABC { let def = ghi(jkl↓:mno) } }"
+            "class ABC { let def = ghi(jkl↓:mno) } }",
+            "func foo() { let dict = [1↓ : 1] }"
         ],
         corrections: [
             "let ↓abc:Void\n": "let abc: Void\n",
@@ -159,7 +161,8 @@ public struct ColonRule: CorrectableRule, ConfigurationProviderRule {
             "object.method(x↓:  5, y: \"string\")\n": "object.method(x: 5, y: \"string\")\n",
             "func abc() { def(ghi↓:jkl) }": "func abc() { def(ghi: jkl) }",
             "func abc(def: Void) { ghi(jkl↓:mno) }": "func abc(def: Void) { ghi(jkl: mno) }",
-            "class ABC { let def = ghi(jkl↓:mno) } }": "class ABC { let def = ghi(jkl: mno) } }"
+            "class ABC { let def = ghi(jkl↓:mno) } }": "class ABC { let def = ghi(jkl: mno) } }",
+            "func foo() { let dict = [1↓ : 1] }": "func foo() { let dict = [1: 1] }"
         ]
     )
 

